ethstorage/scanner/worker.go

```diff
@@ -0,0 +1,214 @@
+	}
+}
+
+func (s *Worker) ScanBatch(ctx context.Context, sendError func(kvIndex uint64, err error)) (*stats, error) {
+	sts := stats{}
+	localKvs, err := s.queryLocalKvs()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get local KV indices: %w", err)
+	}
+	s.lg.Info("Scanner: query local KV entries done", "localKVs", shortPrt(localKvs))
+	sts.total = len(localKvs)
+
+	kvsInBatch, nextKvIndex, err := s.determineBatchIndexRange(localKvs)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get batch index range: %w", err)
+	}
+	metas, err := s.l1.GetKvMetas(kvsInBatch, rpc.LatestBlockNumber.Int64())
+	if err != nil {
+		s.lg.Error("Scanner: error getting meta range", "kvsInBatch", shortPrt(kvsInBatch))
+		return nil, fmt.Errorf("failed to query KV metas %w", err)
+	}
+	s.lg.Info("Scanner: query KV meta done", "kvsInBatch", shortPrt(kvsInBatch))
+	for i, meta := range metas {
+		select {
+		case <-ctx.Done():
+			s.lg.Info("Scanner canceled, stopping scan", "ctx.Err", ctx.Err())
+			return nil, ctx.Err()
+		default:
+		}
+
+		var err error
+		var found bool
+		var commit common.Hash
+		copy(commit[:], meta[32-es.HashSizeInContract:32])
+		kvIndex := kvsInBatch[i]
+		if s.cfg.Mode == modeCheckMeta {
+			// check meta only
+			var metaLocal []byte
+			metaLocal, found, err = s.sm.TryReadMeta(kvIndex)
+			if metaLocal != nil && !bytes.Equal(metaLocal[0:es.HashSizeInContract], commit[0:es.HashSizeInContract]) {
+				err = es.ErrCommitMismatch
+			}
+		} else if s.cfg.Mode == modeCheckBlob {
+			// query blob and check meta from storage
+			_, found, err = s.sm.TryRead(kvIndex, int(s.sm.MaxKvSize()), commit)
+		} else {
+			panic(fmt.Sprintf("invalid scanner mode: %d", s.cfg.Mode))
+		}
+		if found && err == nil {
+			// happy path
+			s.lg.Debug("Scanner: check KV done", "kvIndex", kvIndex, "commit", commit)
+			continue
+		}
+		if !found {
+			// the shard does not exist locally
+			s.lg.Warn("Scanner: blob not found", "kvIndex", kvIndex, "commit", commit)
+			continue
+		}
+		if err != nil {
+			// chances are the blob is downloaded but not written to the storage
+			if blob := s.loadKvFromCache(kvIndex, commit); blob != nil {
+				s.lg.Debug("Scanner: check KV done from cache", "kvIndex", kvIndex, "commit", commit)
+				continue
+			}
+			s.lg.Error("Scanner: read blob error", "kvIndex", kvIndex, "commit", commit.Hex(), "err", err)
+			if err == es.ErrCommitMismatch {
+				sts.mismatched = append(sts.mismatched, kvIndex)
+				if err := s.fixKv(kvIndex, commit); err != nil {
+					sts.failed = append(sts.failed, kvIndex)
+					s.lg.Error("Scanner: fix blob error", "kvIndex", kvIndex, "err", err)
+					sendError(kvIndex, fmt.Errorf("failed to fix blob: %w", err))
+				} else {
+					sts.fixed = append(sts.fixed, kvIndex)
+					// remove the error from the cache
+					sendError(kvIndex, nil)
+				}
+			} else {
+				sendError(kvIndex, fmt.Errorf("failed to read blob: %w", err))
+			}
+		}
+	}
+	s.nextKvIndex = nextKvIndex
```

Comment:
The variable name `nextKvIndex` is misleading since it is actually not the next kv index but the index of next kv index

Reply:
fixed